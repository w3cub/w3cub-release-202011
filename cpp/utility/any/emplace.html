
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Std&#58;&#58;any&#58;&#58;emplace - C++ - W3cubDocs</title>
  
  <meta name="description" content=" Changes the contained object to one of type std&#58;&#58;decay_t&#60;ValueType&#62; constructed from the arguments. ">
  <meta name="keywords" content="std, any, emplace, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/any/emplace.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">std::any::emplace</h1>            <table class="t-dcl-begin"> <tr class="t-dcl t-since-cxx17"> <td> <pre data-language="cpp">template&lt; class ValueType, class... Args &gt;
std::decay_t&lt;ValueType&gt;&amp; emplace( Args&amp;&amp;... args );</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx17">(since C++17)</span> </td> </tr> <tr class="t-dcl t-since-cxx17"> <td> <pre data-language="cpp">template&lt; class ValueType, class U, class... Args &gt;
std::decay_t&lt;ValueType&gt;&amp; emplace( std::initializer_list&lt;U&gt; il, Args&amp;&amp;... args );</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx17">(since C++17)</span> </td> </tr>  </table> <p>Changes the contained object to one of type <code><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;</span></code> constructed from the arguments.</p>
<p>First destroys the current contained object (if any) by <code><a href="reset" title="cpp/utility/any/reset">reset()</a></code>, then:</p>
<span class="t-li">1)</span> constructs an object of type <code><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;</span></code>, <a href="../../language/direct_initialization" title="cpp/language/direct initialization">direct-non-list-initialized</a> from <code><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1121">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>, as the contained object. <ul><li> This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw528">std::<span class="me2">is_constructible_v</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;</span>, Args...<span class="sy1">&gt;</span></code> and <code><a href="http://en.cppreference.com/w/cpp/types/is_copy_constructible"><span class="kw540">std::<span class="me2">is_copy_constructible_v</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;&gt;</span></code> are both <code>true</code>.</li></ul> <span class="t-li">2)</span> constructs an object of type <code><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;</span></code>, <a href="../../language/direct_initialization" title="cpp/language/direct initialization">direct-non-list-initialized</a> from <code>il, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1121">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>, as the contained object. <ul><li> This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw528">std::<span class="me2">is_constructible_v</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;</span>, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args...<span class="sy1">&gt;</span></code> and <code><a href="http://en.cppreference.com/w/cpp/types/is_copy_constructible"><span class="kw540">std::<span class="me2">is_copy_constructible_v</span></span></a><span class="sy1">&lt;</span><a href="http://en.cppreference.com/w/cpp/types/decay"><span class="kw632">std::<span class="me2">decay_t</span></span></a><span class="sy1">&lt;</span>ValueType<span class="sy1">&gt;&gt;</span></code> are both <code>true</code>. </li></ul>  <h3 id="Template_parameters"> Template parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> ValueType </td> <td> - </td> <td> contained value type </td>
</tr> <tr class="t-par-hitem"> <td colspan="3"> Type requirements </td>
</tr> <tr class="t-par-req"> <td colspan="3"> -<code>std::decay_t&lt;ValueType&gt;</code> must meet the requirements of <a href="../../named_req/copyconstructible" title="cpp/named req/CopyConstructible">CopyConstructible</a>. </td>
</tr>
</table> <h3 id="Return_value">  Return value </h3> <p>A reference to the new contained object.</p>
<h3 id="Exceptions">  Exceptions </h3> <p>Throws any exception thrown by <code>T</code>'s constructor. If an exception is thrown, the previously contained object (if any) has been destroyed, and <code>*this</code> does not contain a value.</p>
<h3 id="Example"> Example</h3> <div class="t-example"> <div class="cpp source-cpp"><pre data-language="cpp">#include &lt;algorithm&gt;
#include &lt;any&gt;
#include &lt;iostream&gt;
#include &lt;string&gt;
#include &lt;vector&gt;
 
class Star
{
    std::string name;
    int id;
 
public:
    Star(std::string name, int id) : name{name}, id{id}
    {
        std::cout &lt;&lt; "Star::Star(string, int)\n";
    }
 
    void print() const
    {
        std::cout &lt;&lt; "Star{\"" &lt;&lt; name &lt;&lt; "\" : " &lt;&lt; id &lt;&lt; "};\n";
    }
};
 
int main()
{
    std::any celestial;
    // (1) emplace(Args&amp;&amp;... args);
    celestial.emplace&lt;Star&gt;("Procyon", 2943);
    const auto* star = std::any_cast&lt;Star&gt;(&amp;celestial);
    star-&gt;print();
 
    std::any av;
    // (2) emplace(std::initializer_list&lt;U&gt; il, Args&amp;&amp;... args);
    av.emplace&lt;std::vector&lt;char&gt;&gt;({'C', '+', '+', '1', '7'} /* no args */);
    std::cout &lt;&lt; av.type().name() &lt;&lt; '\n';
    const auto* va = std::any_cast&lt;std::vector&lt;char&gt;&gt;(&amp;av);
    std::for_each(va-&gt;cbegin(), va-&gt;cend(), [](char const&amp; c) { std::cout &lt;&lt; c; });
    std::cout &lt;&lt; '\n';
}</pre></div> <p>Possible output:</p>
<div class="text source-text"><pre data-language="c">Star::Star(string, int)
Star{"Procyon" : 2943};
St6vectorIcSaIcEE
C++17</pre></div> </div> <h3 id="See_also">  See also </h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="any" title="cpp/utility/any/any"> <span class="t-lines"><span>(constructor)</span></span></a></div> </td> <td> constructs an <code>any</code> object <br> <span class="t-mark">(public member function)</span>  </td>
</tr> <tr class="t-dsc"> <td> <div><a href="reset" title="cpp/utility/any/reset"> <span class="t-lines"><span>reset</span></span></a></div> </td> <td> destroys contained object <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    © cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/any/emplace" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/any/emplace</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
