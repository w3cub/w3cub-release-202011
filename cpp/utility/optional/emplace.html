
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Std&#58;&#58;optional&#58;&#58;emplace - C++ - W3cubDocs</title>
  
  <meta name="description" content="Constructs the contained value in-place. If &#42;this already contains a value before the call, the contained value is destroyed by calling its &hellip;">
  <meta name="keywords" content="std, optional, t, emplace, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/optional/emplace.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">std::optional&lt;T&gt;::emplace</h1>            <table class="t-dcl-begin">  <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(1)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class... Args &gt;
T&amp; emplace( Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template&lt; class... Args &gt;
constexpr T&amp; emplace( Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   <tr class="t-dcl-rev-aux"> <td></td> <td rowspan="3">(2)</td> <td></td> </tr> <tr class="t-dcl t-since-cxx17 t-until-cxx20"> <td> <pre data-language="cpp">template&lt; class U, class... Args &gt;
T&amp; emplace( std::initializer_list&lt;U&gt; ilist, Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx17">(since C++17)</span> <br><span class="t-mark-rev t-until-cxx20">(until C++20)</span> </td> </tr> <tr class="t-dcl t-since-cxx20"> <td> <pre data-language="cpp">template&lt; class U, class... Args &gt;
constexpr T&amp; emplace( std::initializer_list&lt;U&gt; ilist, Args&amp;&amp;... args );</pre>
</td>  <td> <span class="t-mark-rev t-since-cxx20">(since C++20)</span> </td> </tr>   </table> <p>Constructs the contained value in-place. If <code>*this</code> already contains a value before the call, the contained value is destroyed by calling its destructor.</p>
<div class="t-li1">
<span class="t-li">1)</span> Initializes the contained value by <a href="../../language/direct_initialization" title="cpp/language/direct initialization">direct-initializing</a> (but not direct-list-initializing) with <code><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1121">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code> as parameters.</div> <div class="t-li1">
<span class="t-li">2)</span> Initializes the contained value by calling its constructor with <code>ilist, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1121">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code> as parameters. This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw527">std::<span class="me2">is_constructible</span></span></a><span class="sy1">&lt;</span>T, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args<span class="sy3">&amp;&amp;</span>...<span class="sy1">&gt;</span><span class="sy4">::</span><span class="me2">value</span></code> is <code>true</code>.</div>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> args... </td> <td> - </td> <td> the arguments to pass to the constructor </td>
</tr> <tr class="t-par"> <td> ilist </td> <td> - </td> <td> the initializer list to pass to the constructor </td>
</tr> <tr class="t-par-hitem"> <td colspan="3"> Type requirements </td>
</tr> <tr class="t-par-req"> <td colspan="3"> -<code>T</code> must be constructible from <code>Args...</code> for overload <span class="t-v">(1)</span> </td>
</tr> <tr class="t-par-req"> <td colspan="3"> -<code>T</code> must be constructible from <code><a href="../initializer_list" title="cpp/utility/initializer list">std::initializer_list</a></code> and <code>Args...</code> for overload <span class="t-v">(2)</span> </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>A reference to the new contained value.</p>
<h3 id="Exceptions"> Exceptions</h3> <p>Any exception thrown by the selected constructor of <code>T</code>. If an exception is thrown, <code>*this</code> does not contain a value after this call (the previously contained value, if any, had been destroyed).</p>
<h3 id="Example"> Example</h3> <div class="t-example"> <div class="cpp source-cpp"><pre data-language="cpp">#include &lt;optional&gt;
#include &lt;iostream&gt;
 
struct A {
    std::string s;
    A(std::string str) : s(std::move(str)), id{n++} { note("+ constructed"); }
    ~A() { note("~ destructed"); }
    A(const A&amp; o) : s(o.s), id{n++} { note("+ copy constructed"); }
    A(A&amp;&amp; o) : s(std::move(o.s)), id{n++} { note("+ move constructed"); }
    A&amp; operator=(const A&amp; other) {
        s = other.s;
        note("= copy assigned");
        return *this;
    }
    A&amp; operator=(A&amp;&amp; other) {
        s = std::move(other.s);
        note("= move assigned");
        return *this;
    }
    inline static int n{};
    int id{};
    void note(auto s) { std::cout &lt;&lt; "  " &lt;&lt; s &lt;&lt; " #" &lt;&lt; id &lt;&lt; '\n'; }
};
 
int main()
{
    std::optional&lt;A&gt; opt;
 
    std::cout &lt;&lt; "Assign:\n";
    opt = A("Lorem ipsum dolor sit amet, consectetur adipiscing elit nec.");
 
    std::cout &lt;&lt; "Emplace:\n";
    // As opt contains a value it will also destroy that value
    opt.emplace("Lorem ipsum dolor sit amet, consectetur efficitur.");
 
    std::cout &lt;&lt; "End example\n";
}</pre></div> <p>Output:</p>
<div class="text source-text"><pre data-language="c">Assign:
  + constructed #0
  + move constructed #1
  ~ destructed #0
Emplace:
  ~ destructed #1
  + constructed #2
End example
  ~ destructed #2</pre></div> </div> <h3 id="Defect_reports"> Defect reports</h3> <p>The following behavior-changing defect reports were applied retroactively to previously published C++ standards.</p>
<table class="dsctable"> <tr> <th> DR </th> <th> Applied to </th> <th> Behavior as published </th> <th> Correct behavior </th>
</tr> <tr> <td> <a rel="nofollow" class="external text" href="https://wg21.link/P2231R1">P2231R1</a> </td> <td> C++20 </td> <td> <code>emplace</code> was not constexpr while the required operations can be constexpr in C++20 </td> <td> made constexpr </td>
</tr>
</table> <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="operator=" title="cpp/utility/optional/operator="> <span class="t-lines"><span>operator=</span></span></a></div> </td> <td> assigns contents <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    © cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/optional/emplace" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/optional/emplace</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
