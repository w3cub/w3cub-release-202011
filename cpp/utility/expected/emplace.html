
<!DOCTYPE HTML>

<html lang="en" class="_theme-default">

<head>
  <meta charset="utf-8">
  <title>Std&#58;&#58;expected&#58;&#58;emplace - C++ - W3cubDocs</title>
  
  <meta name="description" content=" Constructs an expected value in-place. After the call, has_value() returns true. ">
  <meta name="keywords" content="std, expected, t, e, emplace, c++, cpp">
  <meta name="HandheldFriendly" content="True">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="mobile-web-app-capable" content="yes">
  
  <link rel="canonical" href="https://docs.w3cub.com/cpp/utility/expected/emplace.html">
  <link href="/favicon.png" rel="icon">
  <link rel="stylesheet" type="text/css" href="/assets/application-9c7b0d969c0ce639a6c01c2badcdd81667c7c66206aa3d0fccc5cb0585fb1b912dab72ea882101d18010680927514b4b77b107282db8f756fbf63301fe9c88b0.css">
  <script type="text/javascript" src="/assets/application-79c555f6b25481fffac2cac30a7f3e54e608ca09e9e8e42bb1790095ba6d0fcace47d6bc624ddce952c70370892f2d46864f89e6943d4f7f7ff16c8a3231a91a.js"></script>
  <script src="/json/cpp.js"></script>
  
<!-- Google tag (gtag.js) -->
<script async src="https://www.googletagmanager.com/gtag/js?id=G-R3WC07G3GB"></script>
<script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-R3WC07G3GB');
</script>
<script async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js?client=ca-pub-2572770204602497"
     crossorigin="anonymous"></script>
<script async custom-element="amp-auto-ads"
  src="https://cdn.ampproject.org/v0/amp-auto-ads-0.1.js">
</script>


</head>

<body class="docs">
	<amp-auto-ads type="adsense"
              data-ad-client="ca-pub-2572770204602497">
	</amp-auto-ads>
	<div class="_app">
	<header class="_header">

  <a class="_home-link" href="/" ></a>
  <a class="_menu-link"></a>
  <h1 class="_logo">
    <a href="/" class="_nav-link" title="API Documentation Browser">W3cubDocs</a>
  </h1>
  
  <span class="_logo-sub-nav">/</span><span class="_logo-sub-nav"><a href="/cpp/" class="_nav-link" title="" style="margin-left:0;">C++</a></span>
  
  <nav class="_nav">
    <a href="https://tools.w3cub.com/?_sp=docs" target="_blank" class="_nav-link ">W3cubTools</a>
    <a href="/cheatsheets/" class="_nav-link ">Cheatsheets</a>
    <a href="/about/" class="_nav-link ">About</a>
  </nav>
</header>
	<section class="_sidebar">
		
		<form class="_search">
		  <input type="search" class="_search-input" placeholder="Search&hellip;" autocomplete="off" autocapitalize="off" autocorrect="off" spellcheck="false" maxlength="20">
		  <a class="_search-clear"></a>
		  <div class="_search-tag"></div>
		</form>
		
		<div class="_list-wrap">
			<div class="_list">
			
			</div>
		</div>
	</section>
	<section class="_container ">
		<div class="_content">
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="6861657091"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
			<div class="_page _cppref">
				
				
<h1 id="firstHeading" class="firstHeading">std::expected&lt;T,E&gt;::emplace</h1>            <table class="t-dcl-begin"> <tr class="t-dcl-h"> <td> T is not <i>cv</i> void</td> <td></td> <td></td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class... Args &gt;
constexpr T&amp; emplace( Args&amp;&amp;... args ) noexcept;</pre>
</td> <td> (1) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">template&lt; class U, class... Args &gt;
constexpr T&amp; emplace( std::initializer_list&lt;U&gt;&amp; il, Args&amp;&amp;... args ) noexcept;</pre>
</td> <td> (2) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr> <tr class="t-dcl-h"> <td> T is <i>cv</i> void</td> <td></td> <td></td> </tr> <tr class="t-dcl t-since-cxx23"> <td> <pre data-language="cpp">constexpr void emplace() noexcept;</pre>
</td> <td> (3) </td> <td> <span class="t-mark-rev t-since-cxx23">(since C++23)</span> </td> </tr>  </table> <p>Constructs an expected value in-place. After the call, <a href="operator_bool" title="cpp/utility/expected/operator bool"><code>has_value()</code></a> returns true.</p>
<div class="t-li1">
<span class="t-li">1)</span> Destroys the contained value, then initializes the expected value contained in <code>*this</code> as if by <a href="../../language/direct_initialization" title="cpp/language/direct initialization">direct-initializing</a> an object of type <code>T</code> from the arguments <code><a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1121">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>. This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw532">std::<span class="me2">is_nothrow_constructible_v</span></span></a><span class="sy1">&lt;</span>T, Args...<span class="sy1">&gt;</span></code> is true.</div> <div class="t-li1">
<span class="t-li">2)</span> Destroys the contained value, then initializes the expected value contained in <code>*this</code> as if by <a href="../../language/direct_initialization" title="cpp/language/direct initialization">direct-initializing</a> an object of type <code>T</code> from the arguments <code>il, <a href="http://en.cppreference.com/w/cpp/utility/forward"><span class="kw1121">std::<span class="me2">forward</span></span></a><span class="sy1">&lt;</span>Args<span class="sy1">&gt;</span><span class="br0">(</span>args<span class="br0">)</span>...</code>. This overload participates in overload resolution only if <code><a href="http://en.cppreference.com/w/cpp/types/is_constructible"><span class="kw532">std::<span class="me2">is_nothrow_constructible_v</span></span></a><span class="sy1">&lt;</span>T, <a href="http://en.cppreference.com/w/cpp/utility/initializer_list"><span class="kw962">std::<span class="me2">initializer_list</span></span></a><span class="sy1">&lt;</span>U<span class="sy1">&gt;</span><span class="sy3">&amp;</span>, Args...<span class="sy1">&gt;</span></code> is true.</div> <div class="t-li1">
<span class="t-li">3)</span> If <code>*this</code> contains an unexpected value, destroys that value.</div>  <h3 id="Parameters"> Parameters</h3> <table class="t-par-begin"> <tr class="t-par"> <td> args... </td> <td> - </td> <td> the arguments to pass to the constructor </td>
</tr> <tr class="t-par"> <td> il </td> <td> - </td> <td> the initializer list to pass to the constructor </td>
</tr>
</table> <h3 id="Return_value"> Return value</h3> <p>A reference to the new contained value.</p>
<h3 id="Exceptions"> Exceptions</h3> <div class="t-noexcept-inline">
<a href="../../language/noexcept_spec" title="cpp/language/noexcept spec"><code>noexcept</code></a> specification: <div class="t-noexcept-box"><span class="t-cc"><code>noexcept</code></span></div> </div> <h3 id="Notes"> Notes</h3> <p>If the construction of <code>T</code> is potentially-throwing, this function is not defined. In this case, it is the responsibility of the user to create a temporary and move or copy it.</p>
<h3 id="Example"> Example</h3>  <h3 id="See_also"> See also</h3> <table class="t-dsc-begin"> <tr class="t-dsc"> <td> <div><a href="operator=" title="cpp/utility/expected/operator="> <span class="t-lines"><span>operator=</span></span></a></div>
<div><span class="t-lines"><span><span class="t-mark-rev t-since-cxx23">(C++23)</span></span></span></div> </td> <td> assigns contents <br> <span class="t-mark">(public member function)</span>  </td>
</tr> </table>           <div class="_attribution">
  <p class="_attribution-p">
    Â© cppreference.com<br>Licensed under the Creative Commons Attribution-ShareAlike Unported License v3.0.<br>
    <a href="https://en.cppreference.com/w/cpp/utility/expected/emplace" class="_attribution-link">https://en.cppreference.com/w/cpp/utility/expected/emplace</a>
  </p>
</div>

				
			</div>
			<ins class="adsbygoogle"
     style="display:block"
     data-ad-client="ca-pub-2572770204602497"
     data-ad-slot="1992473792"
     data-ad-format="auto"
     data-full-width-responsive="true"></ins>
<script>
     (adsbygoogle = window.adsbygoogle || []).push({});
</script>
		</div>
	</section>

	</div>
	<svg style="display:none">
		<symbol id="icon-dir"><svg viewBox="0 0 20 20"><path d="M15 10c0 .3-.305.515-.305.515l-8.56 5.303c-.625.41-1.135.106-1.135-.67V4.853c0-.777.51-1.078 1.135-.67l8.56 5.305S15 9.702 15 10z"/></svg></symbol>
	  </svg>
</body>
</html>
